botocore/http_session.py

```diff
@@ -0,0 +1,172 @@
+import os.path
+import logging
+import socket
+from base64 import b64encode
+
+from urllib3 import PoolManager, proxy_from_url, Timeout
+from urllib3.exceptions import NewConnectionError, ProtocolError
+
+from botocore.vendored import six
+from botocore.vendored.six.moves.urllib_parse import unquote
+from botocore.awsrequest import AWSResponse
+from botocore.compat import filter_ssl_warnings, urlparse
+from botocore.exceptions import ConnectionClosedError, EndpointConnectionError
+
+try:
+    from urllib3.contrib import pyopenssl
+    pyopenssl.extract_from_urllib3()
+except ImportError:
+    pass
+
+filter_ssl_warnings()
+logger = logging.getLogger(__name__)
+DEFAULT_TIMEOUT = 60
+MAX_POOL_CONNECTIONS = 10
+# TODO: move the vendored cacert when we drop requests
+DEFAULT_CA_BUNDLE = os.path.join(
+    os.path.dirname(__file__), 'vendored', 'requests', 'cacert.pem'
+)
+
+try:
+    from certifi import where
+except ImportError:
+    def where():
+        return DEFAULT_CA_BUNDLE
+
+
+def construct_basic_auth(username, password):
+    auth_str = '{0}:{1}'.format(username, password)
+    encoded_str = b64encode(auth_str.encode('ascii')).strip().decode()
+    return 'Basic {0}'.format(encoded_str)
+
+
+def get_auth_from_url(url):
+    parsed_url = urlparse(url)
+    try:
+        return unquote(parsed_url.username), unquote(parsed_url.password)
+    except (AttributeError, TypeError):
+        return None, None
+
+
+def get_proxy_headers(proxy_url):
+    headers = {}
+    username, password = get_auth_from_url(proxy_url)
+    if username and password:
+        basic_auth = construct_basic_auth(username, password)
+        headers['Proxy-Authorization'] = basic_auth
+    return headers
+
+
+def fix_proxy_url(proxy_url):
+    if proxy_url.startswith('http:') or proxy_url.startswith('https:'):
+        return proxy_url
+    elif proxy_url.startswith('//'):
+        return 'http:' + proxy_url
+    else:
+        return 'http://' + proxy_url
+
+
+def get_cert_path(verify):
+    if verify is not True:
+        return verify
+
+    return where()
+
+
+class Urllib3Session(object):
+    def __init__(self,
+                 verify=True,
+                 proxies=None,
+                 timeout=None,
+                 max_pool_connections=MAX_POOL_CONNECTIONS,
+    ):
+        self._verify = verify
+        self._proxies = proxies or {}
+        if timeout is None:
+            timeout=DEFAULT_TIMEOUT
+        if not isinstance(timeout, (int, float)):
+            timeout = Timeout(connect=timeout[0], read=timeout[1])
+        self._timeout = timeout
+        self._max_pool_connections = max_pool_connections
+        self._proxy_managers = {}
+        self._http_pool = PoolManager(
```

Comment:
What's interesting is that the `max_pool_connections` is really "max persistent connections", and in the case of s3 with virtual hosted addressing, this number is potentially 100 max persistent connections by default (10 per pool, 10 pool total, 1 pool used per bucket because it's a different hostname).  A connection pool would make that closer to what you'd expect `max_pool_connections` to mean, but could potentially result in worse performance.  I suppose we should keep this as is for now to minimize risky changes in the switchover.

Reply:
Yep, that was the big difference I noticed as well.